### PR TITLE
Refactor AR/AP invoices+transactions state handling

### PR DESCRIPTION
**Pull request by ehuelsmann**
**To all reviewing this PR**: please also run this code with some of your regular use-cases. Although the number of changed lines isn't too large, it impacts every action that can be executed for invoices and AR/AP transactions; meaning: the impact surface is huge.

Move state handling to state machines and offer exactly the actions in the UI that are configured in the state machine.

Known issues (to be solved before merge):
* [x] Conversion of a sales order to a sales invoice  
  Observed error: `DBD::Pg::st execute failed: ERROR: invalid input syntax for type integer: "" LINE 1: UPDATE transactions SET workflow_id = '' WHERE id = '59' AND... ^ at /srv/ledgersmb/old/lib/LedgerSMB/IS.pm line 765. '`
* [x] `Delete` action not implemented
* [x] Batch restrictions (within a batch, only posting of the entire batch allowed...) not implemented
* [x] Triggering actions on transactions from batch approval/deletion
* [x] Instantiation of workflows for transactions and invoices created from `AR > Vouchers > AR Voucher` and `AR > Vouchers > Invoice Vouchers`

### CODE REVIEW

**Comment by neilt** [comment]
Is there a description of how to add UI hints ref: [36c3f7f](https://github.com/ledgersmb/LedgerSMB/pull/6876/commits/36c3f7f9a92635b71dbac27b6ed010eff4c04e69)?

File: lib/LedgerSMB/Workflow/Condition/ACL.pm [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  # condition configuration
+  <conditions>
+    <condition name="acl-draft-modify"
+               class="LedgerSMB::Workflow::Condition::ACL"
+               role="draft-modify" />
+  </conditions>
+
+
+=head1 DESCRIPTION
```
[the remark this answers is not in the document]
Comment (ehuelsmann) [reply lib/LedgerSMB/Workflow/Condition/ACL.pm:17]: Done.

File: lib/LedgerSMB/Workflow/Condition/PeriodClosed.pm [hunk 2]
```diff
@@ -0,0 +1,64 @@
+package LedgerSMB::Workflow::Condition::PeriodClosed;
+
+=head1 NAME
+
+LedgerSMB::Workflow::ACL - Workflow condition testing for allowed role
+
+=head1 SYNOPSIS
```
[the remark this answers is not in the document]
Comment (ehuelsmann) [reply lib/LedgerSMB/Workflow/Condition/PeriodClosed.pm:7]: Done.

File: old/bin/aa.pl [hunk 3]
```diff
@@ -662,6 +691,10 @@ sub form_header {
         <th align=right nowrap><label for="ponum">| . $locale->text('PO Number') . qq|</label></th>
         <td><input data-dojo-type="dijit/form/TextBox" name=ponumber id=ponum size=20 value="$form->{ponumber}"></td>
           </tr>
+          <tr>
+          <th align=right nowrap>| . $locale->text('State') . qq|</th>
+          <td>| . ( $wf ? $wf->state : '' ) . qq|</td>
```
Comment (ehuelsmann) [inline old/bin/aa.pl:696]: Translate the state names?

**Comment by ehuelsmann** [comment]
This branch is ready to be tested; where I mean: run manual tests to check that it meets expectations.

**Comment by ehuelsmann** [comment]
This branch is ready to be tested; where I mean: run manual tests to check that it meets expectations.

**Review by ylavoie — APPROVED** [review]
No comments. except for the State translation, which I think should be added if they are to be shown to users.

File: old/bin/aa.pl (hunk 3 above)
Comment (ehuelsmann) [reply old/bin/aa.pl]: Done

**Pull request merged**